Sources/Tracing/Docs.docc/Guides/ImplementATracer.md

```diff
@@ -0,0 +1,186 @@
+
+## Creating an instrument
+
+Creating an instrument means adopting the ``Instrument`` protocol (or ``Tracer`` in case you develop a tracer).
+`Instrument` is part of the `Instrumentation` library & `Tracing` contains the ``Tracer`` protocol.
+
+`Instrument` has two requirements:
+
+1. A method to inject values inside a `FIXME!!!` into a generic carrier (e.g. HTTP headers)
+2. A method to extract values from a generic carrier (e.g. HTTP headers) and store them in a `FIXME!!!`
+
+The two methods will be called by instrumented libraries/frameworks at asynchronous boundaries, giving you a chance to
+act on the provided information or to add additional information to be carried across these boundaries.
+
+> Check out the [`Baggage` documentation](https://github.com/apple/swift-distributed-tracing-baggage) for more information on
+how to retrieve values from the `FIXME!!!` and how to set values on it.
+
+### Creating a `Tracer`
+
+When creating a tracer you need to create two types:
+
+1. Your tracer conforming to ``Tracer``
+2. A span class conforming to ``Span``
+
+> ``Span`` conforms to the standard rules defined in [OpenTelemetry](https://github.com/open-telemetry/opentelemetry-specification/blob/v0.7.0/specification/trace/api.md#span), so if unsure about usage patterns, you can refer to this specification and examples referring to it.
+
+### Defining, injecting and extracting Baggage
+
+```swift
+import Tracing
+
+private enum TraceIDKey: BaggageKey {
+  typealias Value = String
+}
+
+extension Baggage {
+  var traceID: String? {
+    get {
+      return self[TraceIDKey.self]
+    }
+    set {
+      self[TraceIDKey.self] = newValue
+    }
+  }
+}
+
+var context = DefaultFIXME!!!.topLevel(logger: ...)
+context.baggage.traceID = "4bf92f3577b34da6a3ce929d0e0e4736"
+print(context.baggage.traceID ?? "new trace id")
+```
+
+
+### Extracting & injecting Baggage
+
+When hitting boundaries like an outgoing HTTP request you call out to the [configured instrument(s)](#Bootstrapping-the-Instrumentation-System):
+
+An HTTP client e.g. should inject the given `FIXME!!!` into the HTTP headers of its outbound request:
+
+```swift
+func get(url: String, context: FIXME!!!) {
+  var request = HTTPRequest(url: url)
+  InstrumentationSystem.instrument.inject(
+    context.baggage,
+    into: &request.headers,
+    using: HTTPHeadersInjector()
+  )
+}
+```
+
+On the receiving side, an HTTP server should use the following `Instrument` API to extract the HTTP headers of the given
+`HTTPRequest` into:
+
+```swift
+func handler(request: HTTPRequest, context: FIXME!!!) {
+  InstrumentationSystem.instrument.extract(
+    request.headers,
+    into: &context.baggage,
+    using: HTTPHeadersExtractor()
+  )
+  // ...
+}
+```
+
+> In case your library makes use of the `NIOHTTP1.HTTPHeaders` type we already have an `HTTPHeadersInjector` &
+`HTTPHeadersExtractor` available as part of the `NIOInstrumentation` library.
```

Comment:
```suggestion
> In case your library makes use of the `NIOHTTP1.HTTPHeaders` type we already have an `HTTPHeadersInjector` and
`HTTPHeadersExtractor` available as part of the `NIOInstrumentation` library.
```